pept/tracking/birmingham_method/birmingham_method.py

```diff
@@ -0,0 +1,231 @@
+
+		Returns
+		-------
+		self.f_opt	: float
+			A value between 0 and 1 corresponding to the best fopt value for the given dataset. The best value is chosen by
+			minimising the standard deviation of tracked positions over the static dataset. The value may be overwritten using
+			a setter on this class.
+
+		Raises
+		------
+		Exception
+			If 'static_line_data' is not an instance of 'pept.LineData'
+
+		'''
+
+		# Check line_data is an instance (or a subclass!) of pept.LineData
+		if not isinstance(static_line_data, pept.LineData):
+			raise Exception('[ERROR]: line_data should be an instance of pept.LineData')
+
+		self._static_data = static_line_data
+
+		self._static_data.sample_size = sample_size
+
+		std_dev_min = 999
+		f_best = 0
+
+		fig, ax = plt.subplots(1,1,figsize=(5,5))
+
+		for f in fopt:
+			# print(f)
+			locations = []
+			for n in range(1,self._static_data.number_of_samples):
+				LORs = self._static_data.sample_n(n)
+				location, used = birmingham_method(LORs, f)
+				locations.append(location)
+
+			locations = np.array(locations)
+
+			x = locations[:,1]
+			y = locations[:,2]
+			z = locations[:,3]
+
+			err = locations[:,4]
+			
+			std_dev = np.sqrt(x.std()**2 + y.std()**2 + z.std()**2)
+			# ms = (std_dev**2) * (200/(15**2))
+
+			if verbose:
+				print("\nMean positions for fopt = %.2f are: " % f)
+				print("x = %.2f mm +/- %.2f \t  y = %.2f mm +/- %.2f \t z = %.2f mm +/- %.2f" % (x.mean(), x.std(), y.mean(), y.std(), z.mean(), z.std()))
+				print("So the mean error is: %.2f mm" % std_dev)
+				print("The precision of the measurements is: %.2f mm\n" % err.mean())
+
+			accuracy = ax.plot(f,std_dev, c = 'r', marker='o',label="Accuracy",linestyle='')
+			precision = ax.plot(f,err.mean(), c = 'b', marker='s', label="Precision",linestyle='')
+
+			if std_dev < std_dev_min:
+				std_dev_min = std_dev
+				f_best = f
+
+		ax.axvline(f_best,0,50)
+
+		ax.set_xlabel('fopt')
+		ax.set_ylabel('mm')
+
+		ax.set_ylim([0,25])
+
+		red = mlines.Line2D([],[], color='red', marker='o', linestyle='', label = "Accuracy")
+		blue = mlines.Line2D([],[], color='blue', marker='s', linestyle='', label = "Precision")
+
+		plt.legend(handles=[red,blue])
+
+		self._fopt = f_best
+
+	@property
+	def fopt(self):
+		'''The fraction of LORs used to locate a particle
+
+		fopt : instance of pept.LineData
+		'''	
+		return self._fopt
+
+	@fopt.setter
+	def fopt(self, new_fopt):
+		'''The fraction of LORs used to locate a particle
+
+		Parameters
+		----------
+
+		fopt : float between 0 and 1
+			The fraction of original LORs used to produce a final location for a given number of LORs
+
+		Raises
+		------
+		ValueError
+			If 'fopt' is set to be less than 0 or greater than 1.
+		
+		'''
+
+		if (new_fopt > 1) or (new_fopt < 0):
+			raise ValueError('[ERROR]: fopt should be set between 0 and 1')
+
+		self._fopt = new_fopt
+
+
+	def track(self, line_data = None, sample_no=None, sample_size=250, fopt=None, as_array=False):
```

Comment:
Perhaps have a "track_sample" method which uses numpy arrays and a "track" which iterates through some LineData instance, feeding it to "track_sample" in parallel, using joblib.

Reply:
Done!